src/xkernel_core.cpp

```diff
@@ -299,11 +299,13 @@ namespace xeus
     {
         const xjson& content = request.content();
         bool restart = content.value("restart", false);
+        p_interpreter->shutdown_request();
         p_server->stop();
         xjson reply;
         reply["restart"] = restart;
         publish_message("shutdown", xjson::object(), xjson(reply), buffer_sequence());
         send_reply("shutdown_reply", xjson::object(), std::move(reply), c);
+        std::exit(0);
```

Comment:
The call to `std::exit` should be put at the end of the `xserver_zmq::start_impl` method (after the main thread has sent stop requests to other threads).